docs/proposals/hpa-v2.md

```diff
@@ -0,0 +1,294 @@
+}
+
+// a metric describing a Kubernetes object
+type ObjectMetricSource struct {
+    // the described Kubernetes object
+    Target CrossVersionObjectReference `json:"target"`
+
+    // the name of the metric in question
+    MetricName string `json:"metricName"`
+    // the target value of the metric (as a quantity)
+    TargetValue resource.Quantity `json:"targetValue"`
+}
+
+// metric describing pods in the scale target
+type PodsMetricSource struct {
+    // the name of the metric in question
+    MetricName string `json:"metricName"`
+    // the target value of the metric (as a quantity)
+    TargetValue resource.Quantity `json:"targetValue"`
+}
+
+// resource metric describing pods in the scale target
+// (guaranteed to be available and have the same names across clusters)
+type ResourceMetricSource struct {
+    // the name of the resource in question
+    Name api.ResourceName `json:"name"`
+    // the target value of the resource metric as a percentage of the
+    // request on the pods
+    // +optional
+    TargetPercentageOfRequest *int32 `json:"targetPercentageOfRequest,omitempty"`
+    // the target value of the resource metric as a raw value
+    // +optional
+    TargetRawValue resource.Quantity `json:"targetRawValue,omitempty"`
+}
+
+type HorizontalPodAutoscalerStatus struct {
+    // most recent generation observed by this autoscaler.
+    ObservedGeneration *int64 `json:"observedGeneration,omitempty"`
+    // last time the autoscaler scaled the number of pods;
+    // used by the autoscaler to control how often the number of pods is changed.
+    LastScaleTime *unversioned.Time `json:"lastScaleTime,omitempty"`
+
+    // the last observed number of replicas from the target object.
+    CurrentReplicas int32 `json:"currentReplicas"`
+    // the desired number of replicas as last computed by the autoscaler
+    DesiredReplicas int32 `json:"desiredReplicas"`
+
+    // the last read state of the metrics used by this autoscaler
+    CurrentMetrics []MetricStatus `json:"currentMetrics" protobuf:"bytes,5,rep,name=currentMetrics"`
+}
+
+// the status of a single metric
+type MetricStatus struct {
+    // the type of metric source
+    Type MetricSourceType `json:"type"`
+
+    // metric describing a single Kubernetes object
+    Object *ObjectMetricStatus `json:"object,omitemtpy"`
+    // metric describing pods in the scale target
+    Pods *PodsMetricStatus `json:"pods,omitemtpy"`
+    // resource metric describing pods in the scale target
+    Resource *ResourceMetricSource `json:"resource,omitempty"`
+}
+
+// a metric describing a Kubernetes object
+type ObjectMetricStatus struct {
+    // the described Kubernetes object
+    Target CrossVersionObjectReference `json:"target"`
+
+    // the name of the metric in question
+    MetricName string `json:"metricName"`
+    // the current value of the metric (as a quantity)
+    CurrentValue resource.Quantity `json:"targetValue"`
```

Comment:
Maybe I'm missing something, but shouldn't `json:"targetValue"` be `json:"currentValue"` ?